scalyr_agent/copying_manager.py

```diff
@@ -200,6 +200,340 @@ def __init__(self, add_events_request, completion_callback):
         self.next_pipelined_task = None
 
 
+class CopyingManagerFacade(StoppableThread, LogWatcher):
+    """
+    The CopyingManagerFacade is used to hide the complexity of having sharded CopyingManagers.
+
+    It conforms to the same interface as the original CopyingManager, and is responsible for creating
+    and managing multiple copying managers and splitting logs between them.
```

Comment:
In general I think the facade approach is the right way to go with.  Have you given any thought to which methods are needed by the facade version of the CopyManager?  I don't think you need all of the ones below... for example, it won't need a `_send_events` method and a bunch of the private methods.

We'll need to figure out division of responsibility as well.  Does each CopyManager shard have its own checkpoint file that it writes to?  How do we handle that when the number of shards changes? etc

I'm also not a big fan of using `Facade` in the name for this class.. I might go with something like `ShardedCopyManager`.  But, I'm also not going to fight hard on that one.

Reply:
> for example, it won't need a _send_events method and a bunch of the private methods.

Initially it needed them for the tests because I had made CopyingManager private to the module and just exposed CopyingManagerFacade passing through to an internal copying manager and all the tests failed because the tests override CopyingManager.

I'm moving away from that and will just keep the tests for the CopyingManager as is (but using the private class).

I have no preference for using `Facade`, just thought it made the intent clearer in the design if other people read it.  Happy to switch to `ShardedCopyManager` as we both agree on the general approach.

I'm leaving the checkpoints decision until after I've proven whether it works or not.